### PR TITLE
[enterprise-4.6] BZ-1903943: low latency cpu partitioning chapter lacks info

### PR DESCRIPTION
**Pull request by kdrake20**
BZ-1903943: Added multiple partitioning methods

Fixes: BZ-1903943

See https://bugzilla.redhat.com/show_bug.cgi?id=1903943 for additional details.

Preview URL: https://deploy-preview-41100--osdocs.netlify.app/openshift-enterprise/latest/scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes

For release: 4.6 (See PR 41100 for 4.7 and PR [36682](https://github.com/openshift/openshift-docs/pull/36682) for 4.8+)

Signed-off-by: Katie Tothill ktothill@redhat.com

### CODE REVIEW

File: modules/cnf-tuning-nodes-for-low-latency-via-performanceprofile.adoc [hunk 1]
```diff
@@ -60,10 +60,78 @@ spec:
...
 
-Specify two groups of CPUs in the `spec` section:
+[IMPORTANT]
+====
+The reserved and isolated CPU pools must not overlap and together must span all available cores in the worker node.
+====
+
+To ensure that housekeeping tasks and workloads do not interfere with each other, specify two groups of CPUs in the `spec` section of the performance profile. 
+
+* `isolated` - Specifies the CPUs for the application container workloads. These CPUs have the lowest latency. Processes in this group have no interruptions and can, for example, reach much higher DPDK zero packet loss bandwidth.
```
Comment (kquinn1204) [inline modules/cnf-tuning-nodes-for-low-latency-via-performanceprofile.adoc:112]: "Processes in this group have no interruptions and can, for example, reach much higher DPDK zero packet loss bandwidth. " I am having trouble understanding what this sentence is trying to say.

File: scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc [hunk 2]
```diff
@@ -7,17 +8,25 @@ toc::[]
 
 include::modules/cnf-understanding-low-latency.adoc[leveloffset=+1]
 
+.Additional resources
+
+* xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#configuring_hyperthreading_for_a_cluster_{context}[Configuring hyperthreading for a cluster]
```
Comment (tradej) [inline scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc:13]: I believe you can't use `{context}` in an xref, you must provide the actual value for the context variable _of the module you're referring to_ (right now, the context of this assembly is used, and the linked module has a different context). All underscores in the anchor ID also look suspicious.

Comment (tradej) [reply scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc:13]: Same goes for both the other xrefs in the changeset.

File: scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc [hunk 3]
```diff
@@ -7,17 +8,25 @@ toc::[]
...
 
 include::modules/cnf-upgrading-performance-addon-operator.adoc[leveloffset=+1]
 
 include::modules/cnf-provisioning-real-time-and-low-latency-workloads.adoc[leveloffset=+1]
 
-include::modules/cnf-configuring-huge-pages.adoc[leveloffset=+1]
+.Additional resources
 
-include::modules/cnf-allocating-multiple-huge-page-sizes.adoc[leveloffset=+1]
+* xref:../scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc#managing-device-interrupt-processing-for-guaranteed-pod-isolated-cpus_{context}[Managing device interrupt processing for guaranteed pod isolated CPUs]
```
Comment (aireilly) [inline scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc:23]: here too - fix this _{context} to _cnf-master

**Comment by jlojosnegros** [comment]
Content regarding PAO lgtm

**Comment by ktothill** [comment]
Who:
Jose Luis Ojosnegros, SME
Niranjan Mallapadi Raghavender, QE

Date and Time: April 25, 2022  13:19
Summary: Jose and Niranjan provided `lgtm` statuses for both PRs.

Next Steps: Requesting merges from Peer Review Squad.

File: modules/cnf-cpu-infra-container.adoc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+//
+// scalability_and_performance/cnf-performance-addon-operator-for-low-latency-nodes.adoc
+:_content-type: PROCEDURE
+[id="cnf-restricting-cpu-infra-container_{context}"]
+:_context: cnf-master
+= Restricting CPUs for infra and application containers
+
+Generic housekeeping and workload tasks use CPUs in a way that may impact latency-sensitive processes. By default, the container runtime uses all online CPUs to run all containers together, which can result in context switches and spikes in latency. Partitioning the CPUs prevents noisy processes from interfering with latency-sensitive processes by separating them from each other. The following table describes how processes run on a CPU after you have tuned the node using the Performance Addon Operator:
+
+.Process' CPU assignments
```
Comment (kalexand-rh) [inline modules/cnf-cpu-infra-container.adoc:11]: This is too much data to put into the front of a procedure. Please split the module.

File: modules/cnf-cpu-infra-container.adoc [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+|Details
+
+|Burstable and best-effort pods
+|Runs on any CPU except where low latency workload is running
+
+|Infrastructure pods
+|Runs on any CPU except where low latency workload is running
+
+|Interrupts
+|Redirects to reserved CPUs (optional in {product-title} {product-version} and later)
```
Comment (kalexand-rh) [inline modules/cnf-cpu-infra-container.adoc:24]: This sounds like it shouldn't use the {product-version} attribute and instead use 4.6 in both this PR and its companion.

File: modules/cnf-cpu-infra-container.adoc [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+|Pins to reserved CPUs
+
+|Latency-sensitive workload pods
+|Pins to a specific set of exclusive CPUs from the isolated pool
+
+|OS processes/systemd services
+|Pins to reserved CPUs
+|===
+
+The exact partitioning pattern to use depends on many factors like hardware, workload characteristics and the expected system load. Some sample use cases are as follows:
```
Comment (kalexand-rh) [inline modules/cnf-cpu-infra-container.adoc:36]:
```suggestion
The exact partitioning pattern to use depends on many factors like hardware, workload characteristics, and the expected system load. Some sample use cases are as follows:
```

File: modules/cnf-cpu-infra-container.adoc [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+* If the latency-sensitive workload uses specific hardware, such as a network interface card (NIC), ensure that the CPUs in the isolated pool are as close as possible to this hardware. At a minimum, you should place the workload in the same Non-Uniform Memory Access (NUMA) node.
```
Comment (kalexand-rh) [inline modules/cnf-cpu-infra-container.adoc:38]: Per the ISG: 
> ! should verb
Whenever possible, rewrite in a more direct way. Do not use "should" to refer to an action that must be performed.

**Review by kalexand-rh — COMMENTED** [review]
This is close but not quite ready.

**Pull request merged**